### ISSUE TITLE
Support cross-references

### ISSUE THREAD

**Issue by vermiculus** [wallyqs/org-ruby#11]
As you can see by the very ability to link to [this specific heading](https://github.com/vermiculus/ssa-tool/blob/b13f71f3e090de1dabb8ff9c7cbf73942a556c5b/ssa/simulation/Visualization.org#initialization), GitHub's HTML supports auto-linking to headings. Unfortunately, if you try a cross-reference in the linked document, the link is dead. The underlying HTML that is generated looks something like

``` html
<a>My Cross-Reference</a>
```

rather than what (I think) the correct HTML would look like

``` html
<a href="#my-cross-reference">My Cross-Reference</a>
```

Obviously this is a contrived example and the anchor may be different in practice.

I'm not privy to how this functionality is actually effected, so this may not be possible depending on whether these anchors are auto-generated. (Such a generator would be somewhere in github/markup though, yes?) Is this feature feasible?

**Comment by Archenoth** [comment]
Elisp links do something wacky to org-ruby.

I know these aren't supposed to do anything--but instead of nothing, or a broken link, I get a lot of gnarly characters instead.

For example, this Org link "`[[elisp:(switch-to-buffer "*watir*")]["watir"]]`" results in the following being output to HTML: "`&lt;a href=”elisp:(switch-to-buffer “@@html:<b>@@watir@@html:</b>@@”)”&gt;”watir”`".

I have an example of this in [this Gist](https://gist.github.com/Archenoth/9858e98cacee396d0d97#testing).

This is what I see in the above link:
![](https://cloud.githubusercontent.com/assets/1385527/8813288/7f946dbc-2fbf-11e5-95e1-4ad7e6f2175b.png)

**Comment by alphapapa** [comment]
This would be very helpful.  As it is, packages like https://github.com/alphapapa/org-make-toc have to work around this by creating GitHub/`org-ruby`-compatible links manually.  And links made with Org itself don't work.

**Comment by andschwa** [comment]
Just throwing in a vote. I was working on a readme formatting in with `org-mode` and discovered my internal links using the `[[search]]` and `<<direct target>>` formats are broken when rendered on GitHub. This would definitely be nice to have 😄

**Comment by Hurricos** [comment]
I'm also going to take a throw at this, because I really need anchor parsing - at least to HTML - to work for my Gitlab instance.

**Comment by Hurricos** [comment]
A surface-level search suggests that GitLab (https://gitlab.com/gitlab-org/gitlab-markup) and GitHub (https://github.com/github/markup) should be adding anchors after org-ruby is done.  The process involves a lot of sanitization due to the sensitive nature of rendering arbitrary HTML. This does not necessarily explain why GitLab does not provide anchor links next to headers in rendered org-flies, but it suggests that both parsers may be friendly to header links that org-ruby hands to them.

**Comment by Hurricos** [comment]
I looked into it and can confirm that the GitLab/Hub parsers (which come after org-ruby) will always overwrite `id` attributes for the headings. The obvious workaround to actually setting the `id`s - put elements next to the headers with the respective `CUSTOM_ID` and `ID` properties as their HTML `id`s - will also fail because those appear to get scrubbed.

See [this repo](https://github.com/github/markup) for more information about this rendering process.

This can still probably be implemented, though; the standard `id` attribute seems to be made of the header, lower-cased and dash-delimited, so all we really have to do is guess what the header's HTML `id` will look like and set the href to point to that `id`.

**Comment by alphapapa** [comment]
> As @deprecated said, it would be good enough to support just ID and CUSTOM_ID links - org-mode's fuzzy matching is bad practice for so many reasons.

That would not be good enough, because it would require that headings have manually assigned `ID` or `CUSTOM_ID` properties, which is already required to produce links that are followable in both Org and `org-ruby`-generated HTML.  In other words, that is already the case.

What is needed is support for Org's own intra-file heading links.  They are described in the Org manual, in section `4.7 Search options in file links`:
```
   Here is the syntax of the different ways to attach a search to a file
link, together with an explanation:

     [[file:~/code/main.c::255]]
     [[file:~/xx.org::My Target]]
     [[file:~/xx.org::*My Target]]
     [[file:~/xx.org::#my-custom-id]]
     [[file:~/xx.org::/regexp/]]

`255'
     Jump to line 255.

`My Target'
     Search for a link target `<<My Target>>', or do a text search for
     `my target', similar to the search in internal links, see *note
     Internal links::.  In HTML export (*note HTML export::), such a
     file link will become an HTML reference to the corresponding named
     anchor in the linked file.

`*My Target'
     In an Org file, restrict search to headlines.

`#my-custom-id'
     Link to a heading with a `CUSTOM_ID' property

`/regexp/'
     Do a regular expression search for `regexp'.  This uses the Emacs
     command `occur' to list all matches in a separate window.  If the
     target file is in Org mode, `org-occur' is used to create a sparse
     tree with the matches.

   As a degenerate case, a file link with an empty file name can be used
to search the current file.  For example, `[[file:::find me]]' does a
search for `find me' in the current file, just as `[[find me]]' would.
```
Specifically, what's needed is support for these:

For intra-file links, this format:
```
`*My Target'
     In an Org file, restrict search to headlines.
```
Which should be simply expanded to, e.g.:
```html
<a href="#My%20Target">My Target</a>
```
For inter-file links, this format:
```
     [[file:xx.org::*My Target]]
```
Which should be simply expanded to, e.g.:
```html
<a href="xx.org#My%20Target">My Target</a>
```
Where `xx.org` is a file in the same directory.

**Comment by alphapapa** [comment]
> the standard `id` attribute seems to be made of the header, lower-cased and dash-delimited, so all we really have to do is guess what the header's HTML `id` will look like and set the href to point to that `id`. I personally think that's malarkey, but it should work for the most important uses (on GitHub/Lab).

Yes, this is the workaround we have to do in `org-make-toc` and `toc-org`.  And the problem is that it creates links like `[[#heading-name]]` which Org interprets as links to headings with `CUSTOM_ID` properties, so Org can't follow those links unless the user manually assigns such custom IDs to each heading.

It's very frustrating that it's not possible to create links that work both within Org and as rendered by GitHub!

**Comment by vermiculus** [comment]
I don't think manually-assigned `ID` or `CUSTOM_ID` properties aren't worth implementing – even as a separate stepping stone to full support of Org links. The latter may be much harder to do. Making it *possible* to create working intra-document links is worth the inconvenience of additional properties tucked away in a drawer.

**Comment by vermiculus** [comment]
I don't think manually-assigned `ID` or `CUSTOM_ID` properties aren't worth implementing – even as a separate stepping stone to full support of Org links. The latter may be much harder to do. Making it *possible* to create working intra-document links is worth the inconvenience of additional properties tucked away in a drawer.

**Comment by alphapapa** [comment]
@vermiculus Do you mean implementing a command in Emacs to add such ID properties to an Org document automatically, or do you mean something in `org-ruby`?

**Comment by vermiculus** [comment]
Org can already assign IDs to headlines you link to with C-c C-l. It'll generate a GUID to use.

I wouldn't expect org-ruby to make any modifications to the document (ever) nor do any parsing of its generalized link syntax (yet).

**Comment by alphapapa** [comment]
Yes, I meant implementing a command to assign IDs to all headings in a document so they could be linked in a table of contents.

**Comment by vermiculus** [comment]
A generated table of contents is a slightly different request; I'd recommend opening a new issue and marking this one as related. (This is specifically about cross-references created 'manually' (i.e., with a command) that are saved in the documented as committed to git.

**Comment by alphapapa** [comment]
@vermiculus I'm not talking about org-ruby generating a table of contents.  I'm talking about the tables of contents generated by `org-make-toc` and `toc-org`, which generate the ToCs in Emacs using the links and cross references discussed in this issue.  Please see my earlier comments about the problems with such and how org-ruby could be changed to fix them.

**Comment by vermiculus** [comment]
My apologies; I'd been skimming for the phrase 'table of contents', not 'toc' :-) If a table of contents exists literally in an Org document, then of course having that TOC work would be in-scope.

**Comment by ambimorph** [comment]
Just want to echo that it would be extremely useful to have internal links that work both in org-mode and on GitHub.

**Comment by maikol-solis** [comment]
This feature will be super useful to those that keep an org-roam (zettelkasten) repo in github. Many of my links are handled by `org-id` and github cannot parse them at all. 

For example I handle my files like this

```
:PROPERTIES:
:ID:       20200701060753
:END:
#+TITLE: My note
```

And link them like this 

`[[id:20200701060753][My note]]`

**Comment by Timmymayes** [comment]
Adding a vote on this issue.
